### DISCUSSION TITLE
In Openssl 1.1.1t I can see BN_X931_generate_prime_ex is using most CPU. But I can't find the Openssl code invokes it except RSA_X931_generate_key_ex for which I can't find its caller either.

### DISCUSSION THREAD

**Discussion by Syllinia** [openssl/openssl#24050] [Q&A]
Hi Team,

In Openssl 1.1.1t I can see BN_X931_generate_prime_ex is using most CPU. But I can't find the Openssl code invokes it except RSA_X931_generate_key_ex for which I can't find its caller either.  Am I missing something? Or these two functions are called in the form of some macro? Thanks in advance.

**Comment by slontis** [comment]
There are no other calls or macros within OpenSSL.
And RSA_X931_generate_key_ex is only called during tests, in the test file rsa_x391_test.c.

**Reply by Syllinia** [reply]
Thanks for your reply. However, I don't find rsa_x391_test.c in Openssl 1.1.1s.

Besides, in the CPU profiling I do see the usage of BN_X931_generate_prime_ex is ranked highest.

   PerfTop: 2846 irqs/sec kernel:14.5% exact: 100.0% lost: 0/0 drop: 0/0 [4000Hz cycles], (all, CPU: 2)
-------------------------------------------------------------------------------

     7.95% libcrypto.so.1.1 [.] BN_X931_generate_prime_ex

**Reply by t8m** [reply]
Where did you get the OpenSSL build from? Is that a clean upstream build or some system package from an OS distribution?

**Reply by Syllinia** [reply]
I got it from upstream openssl-1.1.1s.tar.gz with my own patches.

**Reply by t8m** [reply]
The question is then what are those your own patches. As @slontis said above these functions are never called from OpenSSL itself. One option is to run the application under debugger and setting a breakpoint on these functions. Then you can display a backtrace of calls that lead to this function.

**Reply by Syllinia** [reply]
@t8m @slontis Thanks so much for your reply.
1.The patches don't call those functions.
2.What's weird, the gdb don't break at these functions.

**Reply by t8m** [reply]
Then PerfTop is perhaps reporting some bogus data?

**Reply by Syllinia** [reply]
My application is a bit complex which uses many shared libraries which may use other shared libraries. So I guess even if my application doesn't call BN_X931_generate_prime_ex directly, if some so calls BN_X931_generate_prime_ex, perf top will count that usage on my application. And both the libraries and my application are stripped off symbol tables, so gdb don't stop on the breakpoints of BN_X931_generate_prime_ex.

**Reply by Syllinia** [reply]
> Then PerfTop is perhaps reporting some bogus data?

Maybe it's the case, because perf top also displays another function __nss_database_lookup in libc. Both __nss_database_lookup and BN_X931_generate_prime_ex doesn't stop in gdb even if I can see them in perf top. Seems BN_X931_generate_prime_ex should only be called in key/certificate generation after I do some research on "X9.31 Prime Derivation".

For __nss_database_lookup, some local symbols which don't have their entries in symbol table and placed after __nss_database_lookup will show themselves as __nss_database_lookup mistakenly.
https://stackoverflow.com/questions/66762523/why-does-nss-database-lookup-take-most-of-the-time-in-my-numerical-c-program
